File: Networkcall/app/src/main/java/com/example/networkcall/CustomAdapter.java
package com.example.networkcall;

import android.view.LayoutInflater;
import android.view.ViewGroup;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;

import com.example.networkcall.CustomViewHolder;
import com.example.networkcall.PojoResponse;
import com.squareup.picasso.Picasso;

public class CustomAdapter
        extends RecyclerView.Adapter<CustomViewHolder> {
    private PojoResponse dataSet;

    public CustomAdapter(PojoResponse dataSet) {
        this.dataSet = dataSet;
    }

    @NonNull
    @Override
    public CustomViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        return new CustomViewHolder(
                LayoutInflater.from(parent.getContext()).inflate(
                        R.layout.item_layout,
                        parent,
                        false
                )
        );
    }

    @Override
    public void onBindViewHolder(@NonNull CustomViewHolder holder, int position) {
        holder.tvItemBookTitle.setText(dataSet.bookList.get(position).volumeInfo.title);
        //todo Picasso implementation
        //todo dataSet.items.get(position).volumeInfo.imageLinks.thumbnail;
        Picasso.get().load(
                dataSet.bookList.get(position).volumeInfo.imageLinks.thumbnail
        ).into(
                holder.ivItemBookCover
        );
    }

    @Override
    public int getItemCount() {
        return dataSet != null ? dataSet.bookList.size() : 0;
    }
}

File: DependencyInjection/app/src/main/java/com/example/countriesdi/CustomAdapter.kt
package com.example.countriesdi

import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ImageView
import android.widget.TextView
import androidx.recyclerview.widget.RecyclerView
import butterknife.BindView
import butterknife.ButterKnife
import com.squareup.picasso.Picasso

class CustomAdapter: RecyclerView.Adapter<CustomAdapter.CustomViewHolder>(){
    var dataSet: List<CountryPoko> = emptyList()
    set(value){
        field = value
        notifyDataSetChanged()
    }
    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): CustomViewHolder {
        return CustomViewHolder(
            LayoutInflater.from(parent.context)
                .inflate(R.layout.item_layout,
                    parent,
                    false)
        )
    }

    override fun getItemCount(): Int {
        return dataSet.size
    }

    override fun onBindViewHolder(holder: CustomViewHolder, position: Int) {
        //bind the data with the viewholder
        holder.countryName.text=dataSet[position].name
        Picasso.get().load(dataSet[position].flag).into(holder.vectorFlag)
        //we need to trigger another network call


    }

    inner class CustomViewHolder(itemView: View) :
        RecyclerView.ViewHolder(itemView) { 
        @BindView(R.id.iv_country_flag)
        lateinit var vectorFlag: ImageView
        @BindView(R.id.tv_country_name)
        lateinit var countryName: TextView
        init{
            ButterKnife.bind(this, itemView)
        }

    }
}